rocAL rocJpeg Integration

fiona-gladwin opened the pull request:
- Introduce rocJpeg hardware decoder in rocAL

spolifroni-amd commented on CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ Full documentation for rocLibrary is available at [https://rocm.docs.amd.com/pro
 ### Added
 * Extended support to rocAL's video decoder to use rocDecode hardware decoder
 * Setup - installs rocdecode dev packages for Ubuntu, RedHat, and SLES
+* Extended support to rocAL's image decoder to use rocJpeg hardware decoder
```
```suggestion
* rocAL's image decoder has been extended to support the rocJPEG hardware decoder
```

spolifroni-amd reviewed (changes requested):
Please review the change I made to the changelog to ensure that the nuance wasn't lost.

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,387 @@
+                    output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = widths[0];
+                    channel_sizes[2] = channel_sizes[1] = channel_sizes[0] = output_image.pitch[0] * heights[0];
+                    break;
+                case ROCJPEG_CSS_422:
```
please add a case for `ROCJPEG_CSS_440`


```
case ROCJPEG_CSS_440:
   num_channels = 3;
   output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = widths[0];
   channel_sizes[0] = output_image.pitch[0] * heights[0];
   channel_sizes[2] = channel_sizes[1] = output_image.pitch[0] * (heights[0]) >> 1);
   break;
```

fiona-gladwin commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,387 @@
+
+    std::string chroma_sub_sampling = "";
+    GetChromaSubsamplingStr(subsampling, chroma_sub_sampling);
+    if (subsampling == ROCJPEG_CSS_440 || subsampling == ROCJPEG_CSS_411 || subsampling == ROCJPEG_CSS_UNKNOWN) {
```
Done

fiona-gladwin commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,387 @@
+        return Status::HEADER_DECODE_FAILED;
+    }
+    *actual_width = scaledw;
+    *actual_height = scaledh;
```
Done

fiona-gladwin commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,387 @@
+    }
+
+    *width = widths[0];
+    *height = heights[0];
```
Done

fiona-gladwin commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,387 @@
+    CHECK_HIP(hipSetDevice(device_id));
+    CHECK_HIP(hipGetDeviceProperties(&hip_dev_prop, device_id));
+
+    std::cout << "Using GPU device " << device_id << ": " << hip_dev_prop.name << "[" << hip_dev_prop.gcnArchName << "] on PCI bus " <<
```
Done

rrawther commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,138 @@
+  unsigned denom;
+};
+
+class HWRocJpegDecoder : public Decoder {
```
rocJpeg is hardware Jpeg decoder. So we don't need to explicity name the class HWRocJpegDecoder. RocJpegDecoder is enough

fiona-gladwin replied:
There is a class with same name present in rocJpeg library over [here](https://github.com/ROCm/rocJPEG/blob/0442180916e0b1b73bdf10a8864db406aad0eb07/src/rocjpeg_decoder.h#L47)

Using the same name for the class in rocAL resulted in segmentation fault while releasing the handle, as the handle class had a shared ptr for this class. Inorder to resolve this seg fault we had renamed the rocJpeg decoder as such

rrawther commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,138 @@
+    int _device_id = 0;
+    hipStream_t _hip_stream;
+
+    // Using the Scaling factors from TurboJpeg decoder
```
Why do we need this in rocJpeg decoder?

fiona-gladwin replied:
When the user passes the max decoded width and height which can be lesser than the image width and height we need to scale down the image.
Hence we have incorporated the scaling method similar to TurboJpeg decoder using it's scaling factors, so it will result in outputs of same dimension to be passed to the dataloader

rrawther reviewed (changes requested):
Added some comments

fiona-gladwin commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+    if (width) *width = widths[0];
+    if (height) *height = heights[0];
+    uint scaledw = widths[0], scaledh = heights[0];
+    // Scaling to be performed if width/height is less than max decode width/height
```
Done

AryanSalmanpour commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,138 @@
+    std::vector<size_t> _src_hstride;
+    std::vector<size_t> _src_img_offset;
+    std::vector<RocJpegImage> _output_images = {};
+    RocJpegDecodeParams _decode_params = {};
```
Please change this line to `std::vector<RocJpegDecodeParams> _decode_params = {};`

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+ * @param channel_sizes The array to store the channel sizes.
+ * @return The channel pitch.
+ */
+int GetChannelPitchAndSizes(RocJpegOutputFormat output_format, RocJpegChromaSubsampling subsampling, uint32_t *widths, uint32_t *heights,
```
Please modify the first argument of the `GetChannelPitchAndSizes` function to accept `RocJpegDecodeParams decode_params` instead of `RocJpegOutputFormat output_format`. You can access the `output_format` using `decode_params.output_format` within the function. This change is necessary because `decode_params` also includes the `crop_rectangle`, which is essential for supporting regions of interest (ROI).

Please take a look to the [GetChannelPitchAndSizes](https://github.com/ROCm/rocJPEG/blob/develop/samples/rocjpeg_samples_utils.h#L318) in rocJPEG and use the same changes here.

fiona-gladwin replied:
Currently we have not added the support in rocAL to decode images with ROI using rocJpeg decoder, Please let me know if this will be required if the ROI is not updated internally.

fiona-gladwin replied:
Done

fiona-gladwin replied:
@AryanSalmanpour added the changes to include ROI in GetChannelPitchAndSizes

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+                case ROCJPEG_CSS_444:
+                    num_channels = 3;
+                    output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = widths[0];
+                    channel_sizes[2] = channel_sizes[1] = channel_sizes[0] = output_image.pitch[0] * heights[0];
```
As I mentioned earlier, please refer to the [GetChannelPitchAndSizes](https://github.com/ROCm/rocJPEG/blob/develop/samples/rocjpeg_samples_utils.h#L318) function in rocJPEG for guidance. You will need to calculate the pitch and channel sizes depending on whether a Region of Interest (ROI) is present. Therefore, the two lines above should be modified accordingly.

```
output_image.pitch[2] = output_image.pitch[1] = output_image.pitch[0] = is_roi_valid ? roi_width : widths[0];
channel_sizes[2] = channel_sizes[1] = channel_sizes[0] = align(output_image.pitch[0] * (is_roi_valid ? roi_height : heights[0]), mem_alignment);

```

AryanSalmanpour replied:
Please see my comment above.

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+        max_heights[0] = (heights[0] + 8) &~ 7;
+    }
+
+    if (GetChannelPitchAndSizes(_decode_params.output_format, subsampling, max_widths, max_heights, channels_size, _output_images[index], channel_sizes)) {
```
Please change `_decode_params.output_format` to `_decode_params[index]`

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+        }
+    }
+
+    CHECK_ROCJPEG(rocJpegDecodeBatched(_rocjpeg_handle, _rocjpeg_streams.data(), _batch_size, &_decode_params, _output_images.data()));
```
As discussed earlier the `_decode_params ` should be changed to a vector so please change the `&_decode_params` to `_decode_params.data()`

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+
+    switch(desired_decoded_color_format) {
+        case Decoder::ColorFormat::GRAY:
+            _decode_params.output_format = ROCJPEG_OUTPUT_Y;
```
please modify the `_decode_params.output_format` to _`decode_params[index].output_format`

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+            break;
+        case Decoder::ColorFormat::RGB:
+        case Decoder::ColorFormat::BGR:
+            _decode_params.output_format = ROCJPEG_OUTPUT_RGB;
```
please modify the _decode_params.output_format to _decode_params[index].output_format

fiona-gladwin replied:
Done

AryanSalmanpour commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+    _output_images.resize(_batch_size);
+    _src_hstride.resize(_batch_size);
+    _src_img_offset.resize(_batch_size);
+
```
Add this line here `_decode_params.resize(_batch_size);` to initialize and resize the `_decode_params` vector.

fiona-gladwin replied:
Done

fiona-gladwin commented:
@AryanSalmanpour addressed all review comments please check

kiritigowda commented:
@fiona-gladwin - Find package for rocJPEG has been updated. Please use the new cmake config files. see rocJPEG samples and tests for how to.

AryanSalmanpour reviewed (changes requested):
@fiona-gladwin in the [rocAL-setup.py](https://github.com/ROCm/rocAL/blob/develop/rocAL-setup.py) add support to install rocjpeg-dev and rocjpeg-devel packages similar to rocdecode-dev/rocdecode-devel as the pre-requisite.

Also in in the [CMakeLists.txt](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt) 
in [line 221](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L221) and [line 222](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L221) add **rocjpeg** after rocdecode 
in [line 223](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L223) add **rocjpeg-dev** after rocdecode-dev
in [line 224](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L224) add **rocjpeg-devel** after rocdecode-devel

rrawther reviewed (approved):
Can be merged after Aryan's comments are resolved

fiona-gladwin commented:
> @fiona-gladwin in the [rocAL-setup.py](https://github.com/ROCm/rocAL/blob/develop/rocAL-setup.py) add support to install rocjpeg-dev and rocjpeg-devel packages similar to rocdecode-dev/rocdecode-devel as the pre-requisite.
> 
> Also in in the [CMakeLists.txt](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt) in [line 221](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L221) and [line 222](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L221) add **rocjpeg** after rocdecode in [line 223](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L223) add **rocjpeg-dev** after rocdecode-dev in [line 224](https://github.com/ROCm/rocAL/blob/develop/CMakeLists.txt#L224) add **rocjpeg-devel** after rocdecode-devel

@AryanSalmanpour I have added the changes in setup.py and CMake [here](https://github.com/ROCm/rocAL/pull/284/commits/88fe64da568dfcf163661d76e790f91fef9e3d9d)
I have also modified the docs [here](https://github.com/ROCm/rocAL/pull/284/commits/eec766e8a50610c26a3e4a23aa7f3e39551653ed)

fiona-gladwin commented:
@LakshmiKumar23  The rocJpeg integration can be tested using the basic test by passing the decoder type as 2 [here](https://github.com/ROCm/rocAL/pull/284/commits/e67809baf69cdb9b184f830addccd8a6e4f4ad24)

fiona-gladwin commented:
@LakshmiKumar23 added ctest support
https://github.com/ROCm/rocAL/pull/284/commits/29a3a19c400a534d44915695c5e198125004906f

LakshmiKumar23 commented on rocAL/source/loaders/image/image_read_and_decode.cpp:
```diff
@@ -279,54 +289,107 @@ ImageReadAndDecode::load(unsigned char *buff,
             _decompressed_buff_ptrs[i] = buff + image_size * i;
 
+        if (_decoder_config._type != DecoderType::ROCJPEG_DEC) {
 #pragma omp parallel for num_threads(_num_threads)  // default(none) TBD: option disabled in Ubuntu 20.04
```
we only mention Ub22+ support. Can remove this comment

fiona-gladwin replied:
Done

kiritigowda commented:
@fiona-gladwin - can you please look at the merge conflicts.

AryanSalmanpour commented on tests/cpp_api/CMakeLists.txt:
```diff
@@ -225,3 +225,14 @@ add_test(
+# 16 - basic_test_rocjpeg -- basic test with rocJpeg
+find_package(HIP QUIET)
+if(HIP_FOUND)
+  if(${HIP_VERSION} VERSION_GREATER_EQUAL 6.5.0)
```
Please enable the basic_test_rocjpeg if rocjpeg is detected and its version exceeds 0.11.0.

```
# add ROCm find Config location
list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/lib/cmake)
find_package(ROCJPEG QUIET)
if(ROCJPEG_FOUND)
     if(${ROCJPEG_VERSION_MAJOR} VERSION_GREATER_EQUAL "0" AND ${ROCJPEG_VERSION_MINOR} VERSION_GREATER_EQUAL "11" AND ${ROCJPEG_VERSION_PATCH} VERSION_GREATER_EQUAL "0")
                  add_test(NAME basic_test_rocjpeg 
                              COMMAND basic_test 
                              ${ROCM_PATH}/share/rocal/test/data/images/AMD-tinyDataSet 
                              ${ROCM_PATH}/share/rocal/test/data/images/AMD-tinyDataSet-val.txt 1 1 224 224 1 1 2
                               WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/basic_test)
    endif()
endif()

```

fiona-gladwin commented:
> @fiona-gladwin - can you please look at the merge conflicts.

Resolved conflicts
